boot/zephyr/main.c

```diff
@@ -371,30 +371,70 @@ void zephyr_boot_log_stop(void)
         */
 
 #if defined(CONFIG_MCUBOOT_SERIAL) || defined(CONFIG_BOOT_USB_DFU_GPIO)
-static bool detect_pin(const char* port, int pin, uint32_t expected, int delay)
+
+#ifdef CONFIG_MCUBOOT_SERIAL
+#define BUTTON_0_DETECT_DELAY CONFIG_BOOT_SERIAL_DETECT_DELAY
+#else
+#define BUTTON_0_DETECT_DELAY CONFIG_BOOT_USB_DFU_DETECT_DELAY
+#endif
+
+
+#define BUTTON_0_NODE DT_ALIAS(mcuboot_button0)
+
+#if DT_NODE_HAS_STATUS(BUTTON_0_NODE, okay) && DT_NODE_HAS_PROP(BUTTON_0_NODE, gpios)
+
+static const struct gpio_dt_spec button0 = GPIO_DT_SPEC_GET(BUTTON_0_NODE, gpios);
+
+#else /* fallback to legacy configuration */
+
+#if defined(CONFIG_MCUBOOT_SERIAL)
+
+#define BUTTON_0_GPIO_LABEL CONFIG_BOOT_SERIAL_DETECT_PORT
+#define BUTTON_0_GPIO_PIN CONFIG_BOOT_SERIAL_DETECT_PIN
+#define BUTTON_0_GPIO_FLAGS ((CONFIG_BOOT_SERIAL_DETECT_PIN_VAL) ?\
+                                (GPIO_ACTIVE_HIGH | GPIO_PULL_DOWN) :\
+                                (GPIO_ACTIVE_LOW | GPIO_PULL_UP))
+
+#elif defined(CONFIG_BOOT_USB_DFU_GPIO)
+
+#define BUTTON_0_GPIO_LABEL CONFIG_BOOT_USB_DFU_DETECT_PORT
+#define BUTTON_0_GPIO_PIN CONFIG_BOOT_USB_DFU_DETECT_PIN
+#define BUTTON_0_GPIO_FLAGS ((CONFIG_BOOT_USB_DFU_DETECT_PIN_VAL) ?\
+                                (GPIO_ACTIVE_HIGH | GPIO_PULL_DOWN) :\
+                                (GPIO_ACTIVE_LOW | GPIO_PULL_UP))
+
+#endif
+
+#define BUTTON_0_LEGACY 1
+
+static struct gpio_dt_spec button0 = {
+	.port = NULL,
+	.pin = BUTTON_0_GPIO_PIN,
+	.dt_flags = BUTTON_0_GPIO_FLAGS
+};
+
+#endif
+
+static bool detect_pin(void)
 {
     int rc;
-    int detect_value;
-    struct device const *detect_port;
+    int pin_active;
 
-    detect_port = device_get_binding(port);
-    __ASSERT(detect_port, "Error: Bad port for boot detection.\n");
+#ifdef BUTTON_0_LEGACY
+    button0.port = device_get_binding(BUTTON_0_GPIO_LABEL);
+    __ASSERT(button0.port, "Error: Bad port for boot detection.\n");
+#endif
```

Comment:
For non-legacy setup, `device_is_ready(button0.port)` should be called before `gpio_pin_configure_dt(&button0, ...)`.

Reply:
fixed